flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultExecutionDeployer.java

```diff
@@ -0,0 +1,386 @@
+        } else {
+            return failure;
+        }
+    }
+
+    private Function<LogicalSlot, CompletableFuture<Void>> registerProducedPartitions(
+            final ExecutionDeploymentHandle deploymentHandle) {
+
+        return logicalSlot -> {
+            // a null logicalSlot means the slot assignment is skipped, in which case
+            // the produced partition registration process can be skipped as well
+            if (logicalSlot != null) {
+                final Execution execution = getExecution(deploymentHandle.getExecutionAttemptId());
+                final CompletableFuture<Void> partitionRegistrationFuture =
+                        execution.registerProducedPartitions(logicalSlot.getTaskManagerLocation());
+
+                return FutureUtils.orTimeout(
+                        partitionRegistrationFuture,
+                        partitionRegistrationTimeout.toMilliseconds(),
+                        TimeUnit.MILLISECONDS,
+                        mainThreadExecutor);
+            } else {
+                return FutureUtils.completedVoidFuture();
+            }
+        };
+    }
+
+    private BiFunction<Object, Throwable, Void> deployOrHandleError(
+            final ExecutionDeploymentHandle deploymentHandle) {
+
+        return (ignored, throwable) -> {
+            final ExecutionVertexVersion requiredVertexVersion =
+                    deploymentHandle.getRequiredVertexVersion();
+            final Execution execution = getExecution(deploymentHandle.getExecutionAttemptId());
+
+            if (execution == null
+                    || execution.getState() != ExecutionState.SCHEDULED
+                    || executionVertexVersioner.isModified(requiredVertexVersion)) {
+                if (throwable == null) {
+                    log.debug(
+                            "Refusing to assign slot to execution {} because this deployment was "
+                                    + "superseded by another deployment",
+                            deploymentHandle.getExecutionAttemptId());
+                }
+                return null;
+            }
+
+            if (throwable == null) {
+                deployTaskSafe(execution);
+            } else {
+                handleTaskDeploymentFailure(execution.getAttemptId(), throwable);
+            }
+            return null;
+        };
+    }
+
+    private void deployTaskSafe(final Execution execution) {
+        try {
+            executionOperations.deploy(execution);
+        } catch (Throwable e) {
+            handleTaskDeploymentFailure(execution.getAttemptId(), e);
+        }
+    }
+
+    private void handleTaskDeploymentFailure(
+            final ExecutionAttemptID executionAttemptId, final Throwable error) {
+
+        final Execution execution = getExecution(executionAttemptId);
+        executionOperations.markFailed(execution, error);
+    }
+
+    private Execution getExecution(ExecutionAttemptID executionAttemptId) {
```

Comment:
Can the return value be `null` ? If yes, maybe add `@Nullable`  and add checks.

Reply:
Changed the return type to `Optional<Execution>` (Flink code style suggests the return type to be `Optional` instead of `Nullable`). A `getExecutionorThrow(...)` is added which alway expects the execution to be non-null.